File: 139.单词拆分.py
#
# @lc app=leetcode.cn id=139 lang=python3
#
# [139] 单词拆分
#

# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        word_dict = {}
        for word in wordDict:   # 字典加速访问
            word_dict[word] = True
        dp = [False for _ in range(len(s)+1)]
        dp[0] = True

        for i in range(1, len(s) + 1):
            for j in range(i):
                # s[:j] 存在同时 s[j:i] 存在，表示当前部分可以由字典中字符串组成
                if dp[j] and s[j:i] in word_dict:
                    dp[i] = True
                    break
        return dp[len(s)]
# @lc code=end

File: 139.单词拆分.go
/*
 * @lc app=leetcode.cn id=139 lang=golang
 *
 * [139] 单词拆分
 */

// @lc code=start
func wordBreak(s string, wordDict []string) bool {
    wordDictSet := make(map[string]bool)
    for _, w := range wordDict {    // 转为字典加速访问
        wordDictSet[w] = true
    }
    dp := make([]bool, len(s) + 1)
    dp[0] = true	// 空字符串也是组成部分
    for i := 1; i <= len(s); i++ {
        for j := 0; j < i; j++ {
            if dp[j] && wordDictSet[s[j:i]] {   // s[:j] 存在同时 s[j:i] 存在
                dp[i] = true
                break
            }
        }
    }
    return dp[len(s)]
}

// @lc code=end

